File: backend/src/config/aws-secrets.ts
/**
 * AWS Secrets Manager Integration
 * 
 * This module provides integration with AWS Secrets Manager
 * for secure configuration management.
 */

import {
  SecretsManagerClient,
  GetSecretValueCommand,
  GetSecretValueCommandOutput
} from '@aws-sdk/client-secrets-manager';

export class AWSSecretsManager {
  private client: SecretsManagerClient;
  private cache: Map<string, { value: any; timestamp: number }> = new Map();
  private readonly CACHE_TTL = 5 * 60 * 1000; // 5 minutes

  constructor(region: string = process.env.AWS_REGION || 'us-east-1') {
    this.client = new SecretsManagerClient({ region });
  }

  /**
   * Get secret value from AWS Secrets Manager
   */
  async getSecret(secretName: string): Promise<Record<string, any>> {
    // Check cache first
    const cached = this.cache.get(secretName);
    if (cached && Date.now() - cached.timestamp < this.CACHE_TTL) {
      console.log(`📦 Using cached secret: ${secretName}`);
      return cached.value;
    }

    try {
      console.log(`🔐 Fetching secret: ${secretName}`);
      
      const command = new GetSecretValueCommand({
        SecretId: secretName,
        VersionStage: 'AWSCURRENT'
      });

      const response: GetSecretValueCommandOutput = await this.client.send(command);
      
      let secretValue: Record<string, any>;
      
      if (response.SecretString) {
        secretValue = JSON.parse(response.SecretString);
      } else if (response.SecretBinary) {
        const buff = Buffer.from(response.SecretBinary);
        secretValue = JSON.parse(buff.toString('utf-8'));
      } else {
        throw new Error('Secret value is empty');
      }

      // Cache the secret
      this.cache.set(secretName, {
        value: secretValue,
        timestamp: Date.now()
      });

      console.log(`✅ Successfully retrieved secret: ${secretName}`);
      return secretValue;
    } catch (error: any) {
      console.error(`❌ Error retrieving secret ${secretName}:`, error.message);
      
      // Fallback to environment variables if Secrets Manager fails
      if (error.name === 'ResourceNotFoundException') {
        console.log('⚠️ Secret not found, falling back to environment variables');
        return this.getEnvFallback();
      }
      
      throw error;
    }
  }

  /**
   * Get specific secret value
   */
  async getSecretValue(secretName: string, key: string): Promise<string | undefined> {
    const secrets = await this.getSecret(secretName);
    return secrets[key];
  }

  /**
   * Load secrets into process.env
   */
  async loadSecretsToEnv(secretName: string): Promise<void> {
    try {
      const secrets = await this.getSecret(secretName);
      
      Object.entries(secrets).forEach(([key, value]) => {
        if (value && !process.env[key]) {
          process.env[key] = String(value);
        }
      });

      console.log('✅ Secrets loaded into environment variables');
    } catch (error) {
      console.error('❌ Failed to load secrets to environment:', error);
      throw error;
    }
  }

  /**
   * Clear cache
   */
  clearCache(): void {
    this.cache.clear();
    console.log('🧹 Secret cache cleared');
  }

  /**
   * Fallback to environment variables
   */
  private getEnvFallback(): Record<string, any> {
    return {
      DB_HOST: process.env.DB_HOST || 'localhost',
      DB_PORT: process.env.DB_PORT || '5432',
      DB_NAME: process.env.DB_NAME || 't_developer',
      DB_USER: process.env.DB_USER || 'postgres',
      DB_PASSWORD: process.env.DB_PASSWORD || '',
      REDIS_HOST: process.env.REDIS_HOST || 'localhost',
      REDIS_PORT: process.env.REDIS_PORT || '6379',
      REDIS_PASSWORD: process.env.REDIS_PASSWORD || '',
      JWT_SECRET: process.env.JWT_SECRET || 'dev-secret-key',
      AWS_REGION: process.env.AWS_REGION || 'us-east-1',
      BEDROCK_AGENT_ID: process.env.BEDROCK_AGENT_ID || '',
      BEDROCK_AGENT_ALIAS_ID: process.env.BEDROCK_AGENT_ALIAS_ID || ''
    };
  }
}

/**
 * Configuration manager that uses AWS Secrets Manager
 */
export class ConfigManager {
  private static instance: ConfigManager;
  private secretsManager: AWSSecretsManager;
  private environment: string;
  private config: Record<string, any> = {};

  private constructor() {
    this.secretsManager = new AWSSecretsManager();
    this.environment = process.env.NODE_ENV || 'development';
  }

  static getInstance(): ConfigManager {
    if (!ConfigManager.instance) {
      ConfigManager.instance = new ConfigManager();
    }
    return ConfigManager.instance;
  }

  /**
   * Initialize configuration from AWS Secrets Manager
   */
  async initialize(): Promise<void> {
    const secretName = `t-developer/${this.environment === 'production' ? 'prod' : this.environment}`;
    
    try {
      // Load secrets from AWS Secrets Manager
      await this.secretsManager.loadSecretsToEnv(secretName);
      
      // Build configuration object
      this.config = {
        app: {
          name: 'T-Developer MVP',
          environment: this.environment,
          port: parseInt(process.env.PORT || '8000'),
          apiPrefix: '/api/v1'
        },
        database: {
          host: process.env.DB_HOST,
          port: parseInt(process.env.DB_PORT || '5432'),
          name: process.env.DB_NAME,
          user: process.env.DB_USER,
          password: process.env.DB_PASSWORD,
          ssl: this.environment === 'production'
        },
        redis: {
          host: process.env.REDIS_HOST,
          port: parseInt(process.env.REDIS_PORT || '6379'),
          password: process.env.REDIS_PASSWORD
        },
        auth: {
          jwtSecret: process.env.JWT_SECRET,
          jwtExpiresIn: process.env.JWT_EXPIRES_IN || '7d',
          bcryptRounds: 10
        },
        aws: {
          region: process.env.AWS_REGION || 'us-east-1',
          accessKeyId: process.env.AWS_ACCESS_KEY_ID,
          secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
          bedrock: {
            agentId: process.env.BEDROCK_AGENT_ID,
            agentAliasId: process.env.BEDROCK_AGENT_ALIAS_ID,
            modelId: process.env.BEDROCK_MODEL_ID || 'anthropic.claude-v2'
          }
        },
        oauth: {
          google: {
            clientId: process.env.GOOGLE_CLIENT_ID,
            clientSecret: process.env.GOOGLE_CLIENT_SECRET
          },
          github: {
            clientId: process.env.GITHUB_CLIENT_ID,
            clientSecret: process.env.GITHUB_CLIENT_SECRET
          }
        },
        monitoring: {
          datadogApiKey: process.env.DATADOG_API_KEY,
          sentryDsn: process.env.SENTRY_DSN
        },
        email: {
          smtp: {
            host: process.env.SMTP_HOST,
            port: parseInt(process.env.SMTP_PORT || '587'),
            user: process.env.SMTP_USER,
            password: process.env.SMTP_PASSWORD
          },
          sendgridApiKey: process.env.SENDGRID_API_KEY
        },
        features: {
          enableCache: process.env.ENABLE_CACHE === 'true',
          enableMetrics: process.env.ENABLE_METRICS === 'true',
          enableTracing: process.env.ENABLE_TRACING === 'true'
        }
      };

      console.log(`✅ Configuration initialized for ${this.environment} environment`);
    } catch (error) {
      console.error('❌ Failed to initialize configuration:', error);
      
      // Use fallback configuration
      this.config = this.getFallbackConfig();
      console.log('⚠️ Using fallback configuration');
    }
  }

  /**
   * Get configuration value
   */
  get<T = any>(path: string): T {
    const keys = path.split('.');
    let value: any = this.config;
    
    for (const key of keys) {
      value = value?.[key];
      if (value === undefined) break;
    }
    
    return value as T;
  }

  /**
   * Get entire configuration
   */
  getAll(): Record<string, any> {
    return { ...this.config };
  }

  /**
   * Refresh configuration from Secrets Manager
   */
  async refresh(): Promise<void> {
    this.secretsManager.clearCache();
    await this.initialize();
    console.log('🔄 Configuration refreshed');
  }

  /**
   * Get fallback configuration
   */
  private getFallbackConfig(): Record<string, any> {
    return {
      app: {
        name: 'T-Developer MVP',
        environment: this.environment,
        port: 8000,
        apiPrefix: '/api/v1'
      },
      database: {
        host: 'localhost',
        port: 5432,
        name: 't_developer',
        user: 'postgres',
        password: 'postgres',
        ssl: false
      },
      redis: {
        host: 'localhost',
        port: 6379,
        password: ''
      },
      auth: {
        jwtSecret: 'dev-secret-key',
        jwtExpiresIn: '7d',
        bcryptRounds: 10
      },
      aws: {
        region: 'us-east-1',
        bedrock: {
          modelId: 'anthropic.claude-v2'
        }
      },
      features: {
        enableCache: true,
        enableMetrics: false,
        enableTracing: false
      }
    };
  }
}

// Export singleton instance
export const configManager = ConfigManager.getInstance();

File: backend/src/config/aws_secrets.py
"""
AWS Secrets Manager and Parameter Store integration
Securely loads API keys and configuration from AWS services
"""
import json
import logging
import os
from typing import Any, Dict, Optional

import boto3
from botocore.exceptions import ClientError

logger = logging.getLogger(__name__)


class AWSSecretsManager:
    """Manages secrets from AWS Secrets Manager and Parameter Store"""

    def __init__(self, region: str = "us-east-1"):
        self.region = region
        self.secrets_client = boto3.client("secretsmanager", region_name=region)
        self.ssm_client = boto3.client("ssm", region_name=region)
        self.environment = os.getenv("ENVIRONMENT", "development")
        self.secrets_cache = {}
        self.parameters_cache = {}

    def get_secret(self, secret_name: str) -> Optional[Dict[str, Any]]:
        """Get secret from AWS Secrets Manager"""

        # Check cache
        if secret_name in self.secrets_cache:
            return self.secrets_cache[secret_name]

        # Full secret name with environment
        full_secret_name = f"t-developer/{self.environment}/{secret_name}"

        try:
            response = self.secrets_client.get_secret_value(SecretId=full_secret_name)

            # Parse secret string
            if "SecretString" in response:
                secret = json.loads(response["SecretString"])
                self.secrets_cache[secret_name] = secret
                logger.info(f"✅ Loaded secret: {secret_name}")
                return secret
            else:
                # Binary secret
                return {"binary": response["SecretBinary"]}

        except ClientError as e:
            if e.response["Error"]["Code"] == "ResourceNotFoundException":
                logger.warning(f"Secret not found: {full_secret_name}")
            else:
                logger.error(f"Error retrieving secret {full_secret_name}: {e}")
            return None
        except Exception as e:
            logger.error(f"Unexpected error retrieving secret: {e}")
            return None

    def get_parameter(self, parameter_name: str, decrypt: bool = True) -> Optional[str]:
        """Get parameter from AWS Systems Manager Parameter Store"""

        # Check cache
        if parameter_name in self.parameters_cache:
            return self.parameters_cache[parameter_name]

        # Full parameter name with environment
        full_parameter_name = f"/t-developer/{self.environment}/{parameter_name}"

        try:
            response = self.ssm_client.get_parameter(
                Name=full_parameter_name, WithDecryption=decrypt
            )

            value = response["Parameter"]["Value"]
            self.parameters_cache[parameter_name] = value
            logger.info(f"✅ Loaded parameter: {parameter_name}")
            return value

        except ClientError as e:
            if e.response["Error"]["Code"] == "ParameterNotFound":
                logger.warning(f"Parameter not found: {full_parameter_name}")
            else:
                logger.error(f"Error retrieving parameter {full_parameter_name}: {e}")
            return None
        except Exception as e:
            logger.error(f"Unexpected error retrieving parameter: {e}")
            return None

    def get_parameters_by_path(self, path: str, decrypt: bool = True) -> Dict[str, str]:
        """Get all parameters under a path"""

        full_path = f"/t-developer/{self.environment}/{path}"
        parameters = {}

        try:
            paginator = self.ssm_client.get_paginator("get_parameters_by_path")
            pages = paginator.paginate(Path=full_path, Recursive=True, WithDecryption=decrypt)

            for page in pages:
                for parameter in page["Parameters"]:
                    # Extract parameter name without path
                    name = parameter["Name"].replace(full_path + "/", "")
                    parameters[name] = parameter["Value"]

            logger.info(f"✅ Loaded {len(parameters)} parameters from path: {path}")
            return parameters

        except Exception as e:
            logger.error(f"Error retrieving parameters by path {full_path}: {e}")
            return {}

    def set_parameter(self, parameter_name: str, value: str, secure: bool = True) -> bool:
        """Set parameter in Parameter Store"""

        full_parameter_name = f"/t-developer/{self.environment}/{parameter_name}"

        try:
            self.ssm_client.put_parameter(
                Name=full_parameter_name,
                Value=value,
                Type="SecureString" if secure else "String",
                Overwrite=True,
            )

            # Update cache
            self.parameters_cache[parameter_name] = value
            logger.info(f"✅ Set parameter: {parameter_name}")
            return True

        except Exception as e:
            logger.error(f"Error setting parameter {full_parameter_name}: {e}")
            return False

    def set_secret(self, secret_name: str, secret_value: Dict[str, Any]) -> bool:
        """Set secret in Secrets Manager"""

        full_secret_name = f"t-developer/{self.environment}/{secret_name}"

        try:
            # Try to update existing secret
            try:
                self.secrets_client.update_secret(
                    SecretId=full_secret_name, SecretString=json.dumps(secret_value)
                )
            except ClientError as e:
                if e.response["Error"]["Code"] == "ResourceNotFoundException":
                    # Create new secret
                    self.secrets_client.create_secret(
                        Name=full_secret_name, SecretString=json.dumps(secret_value)
                    )
                else:
                    raise e

            # Update cache
            self.secrets_cache[secret_name] = secret_value
            logger.info(f"✅ Set secret: {secret_name}")
            return True

        except Exception as e:
            logger.error(f"Error setting secret {full_secret_name}: {e}")
            return False

    def load_ai_credentials(self) -> Dict[str, Any]:
        """Load all AI provider credentials"""

        credentials = {}

        # Load from Secrets Manager (sensitive data)
        ai_secrets = self.get_secret("ai-credentials")
        if ai_secrets:
            credentials.update(ai_secrets)

        # Individual API keys from Secrets Manager
        openai_key = self.get_secret("openai-api-key")
        if openai_key and isinstance(openai_key, dict):
            credentials["OPENAI_API_KEY"] = openai_key.get("api_key")

        anthropic_key = self.get_secret("anthropic-api-key")
        if anthropic_key and isinstance(anthropic_key, dict):
            credentials["ANTHROPIC_API_KEY"] = anthropic_key.get("api_key")

        # Load from Parameter Store (configuration)
        ai_config = self.get_parameters_by_path("ai-config")
        credentials.update(ai_config)

        # Individual parameters
        params = [
            "openai_model",
            "anthropic_model",
            "bedrock_model_id",
            "ai_provider",
            "ai_temperature",
            "ai_max_tokens",
        ]

        for param in params:
            value = self.get_parameter(param)
            if value:
                credentials[param.upper()] = value

        return credentials

    def initialize_environment(self) -> Dict[str, str]:
        """Initialize environment variables from AWS"""

        env_vars = {}

        # Load AI credentials
        ai_creds = self.load_ai_credentials()
        for key, value in ai_creds.items():
            if value:
                os.environ[key] = str(value)
                env_vars[key] = str(value)

        # Load general configuration from Parameter Store
        config_params = self.get_parameters_by_path("config")
        for key, value in config_params.items():
            env_key = key.upper().replace("-", "_")
            os.environ[env_key] = value
            env_vars[env_key] = value

        logger.info(f"✅ Loaded {len(env_vars)} environment variables from AWS")
        return env_vars


# Global instance
aws_secrets = AWSSecretsManager()

# Initialize on import
try:
    aws_secrets.initialize_environment()
    logger.info("AWS Secrets Manager initialized and environment loaded")
except Exception as e:
    logger.warning(f"Failed to initialize AWS Secrets Manager: {e}")
